docs/hugo/content/design/ADR-2022-11-Change-Detection.md

```diff
@@ -4,79 +4,115 @@ title: '2022-11: Change Detection'
+### Issue: Default values
 
-While we can tell the different in a Spec between a field that was never set by the user, and a field that we set to an explicit value (because all our Spec properties are pointers), this doesn't mean we can assume any explicit value in a Status is a default. For example, if a third party makes a change via the Azure portal, we want to detect this as a change and and revert it.
+Some fields are optional in the Swagger specification but required by Azure. In these cases, Azure will set a default value if the user does not provide one. In a naive implementation, we'll see these fields as differences between the _Spec_ we have and the _Status_ we retrieve from Azure. Complicating this further, some resources may use different default values depending on other fields. (We know this is the case for AKS, for example.)
+
+While we can tell the difference in a Spec between a field that was never set by the user and a field that was set to an explicit value (because all our Spec properties are pointers), this doesn't mean we can assume any explicit value in a Status is a default. For example, if a third party makes a change via the Azure portal, we want to detect this as a change and revert it.
 
 A variation of value defaulting may occur with arrays and maps, where some services will add a default value to the collection if the user does not specify it. Conceptually, you can write `["a"]` and get back `["default", "a"]` when you read it back.
 
-### Read-After-Write Consistency
+**Potential mitigation**: Keep track of the changes made by Azure when we originally PUT the resource, and use those as another input to our change detection process. For most resources, the current state of the resource is returned in the response to either the PUT or the long-running-operation (LRO) polling operation, so we can capture these changes at that point. For others, we may need to perform an initial GET after the PUT to capture these changes, despite the inherent race condition of doing so.
+
+### Issue: Read-After-Write Consistency
 
-Azure is not guaranteed to return the exact same resource that was PUT. 
+Azure is not guaranteed to return the exact same resource that was PUT.
 
 * Some services normalize identifier fields (converting them to all lowercase)
 * Some fields accept a wider range of values than they return.
 
-For example the CosmosDB API will return the region `West US` when the resource specified `westus`.
+For example, the CosmosDB API will return the region `West US` when the resource specified `westus`.
+
+**Potential mitigations**:
+
+* Generally speaking, Azure identifiers are expected to be "case-preserving, case-insensitive", so we should compare those string fields in a case-insensitive manner.
```

Comment:
Only required for IDs (names, etc). Enum values and/or regions as we call out (also VM SKUs, which are somewhat enum-y) are all supposed to not be normalized but sometimes are.

tl;dr: I think we'll be playing whack-a-mole with fields if our approach doesn't handle them in a holistic fashion.